src/daft-plan/src/treenode.rs

```diff
@@ -0,0 +1,85 @@
+use std::sync::Arc;
+
+use crate::{physical_plan::PhysicalPlan, LogicalPlan};
+use common_error::DaftResult;
+use common_treenode::{TreeNode, VisitRecursion};
+
+impl TreeNode for LogicalPlan {
+    fn apply_children<F>(&self, op: &mut F) -> DaftResult<common_treenode::VisitRecursion>
+    where
+        F: FnMut(&Self) -> DaftResult<common_treenode::VisitRecursion>,
+    {
+        for child in self.children().into_iter() {
+            match op(child.as_ref())? {
+                VisitRecursion::Continue => {}
+                VisitRecursion::Skip => return Ok(VisitRecursion::Continue),
+                VisitRecursion::Stop => return Ok(VisitRecursion::Stop),
+            }
+        }
+        Ok(VisitRecursion::Continue)
+    }
+
+    fn map_children<F>(self, transform: F) -> DaftResult<Self>
+    where
+        F: FnMut(Self) -> DaftResult<Self>,
+    {
+        let mut transform = transform;
+        let old_children = self.children();
+        let new_children = old_children
+            .iter()
+            .map(|c| transform(c.as_ref().clone()).map(Arc::new))
+            .collect::<DaftResult<Vec<_>>>()?;
+
+        if old_children
```

Comment:
We might not need this optimization for the common case. For the more complex cases I think we would use a `TreeNodeRewriter` which checks for a Mutate enum value
https://github.com/Eventual-Inc/Daft/blob/1fcb70ef5fd64b194feab64de3a984a24c62cfe9/src/common/treenode/src/lib.rs#L195